### PR TITLE
Presentation: Add support for custom property category renderers

### PR DESCRIPTION
**Pull request by roluk**
- Extends `PropertyCategorySpecification` to include a `renderer` field.
- Adds `PropertyCategoryRendererManager` which acts similarly to `PropertyValueRendererManager`, but is specific to property grid categories.
- Integrates custom category rendering logic into `VirtualizedPropertyGrid`.
- Some previously internal property grid state and types are now exposed to custom category renderers.
- Documents custom property category renderer usage and adds entry to `NextVersion.md`

[Parallel PR for the native addon.](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/155584)

### CODE REVIEW

File: ui/components/src/ui-components/propertygrid/component/VirtualizedPropertyGridWithDataProvider.tsx [hunk 1]
```diff
@@ -22,7 +23,10 @@ import { VirtualizedPropertyGrid } from "./VirtualizedPropertyGrid";
  */
 export interface VirtualizedPropertyGridWithDataProviderProps extends CommonPropertyGridProps {
   dataProvider: IPropertyDataProvider;
-  highlight?: HighlightingComponentProps & {filteredTypes?: FilteredType[]};
+  highlight?: HighlightingComponentProps & {
+    filteredTypes?: FilteredType[];
+  };
+  propertyCategoryRendererManager?: PropertyCategoryRendererManager;
```
Comment (grigasp) [inline ui/components/src/ui-components/propertygrid/component/VirtualizedPropertyGridWithDataProvider.tsx:29]: I think should be `@alpha`?

File: ui/components/src/ui-components/propertygrid/component/VirtualizedPropertyGrid.tsx [hunk 2]
```diff
@@ -33,14 +34,18 @@ import { FlatItemNestedBorderWrapper } from "./FlatItemNestedBorderWrapper";
...
  * @alpha
  */
 export interface VirtualizedPropertyGridProps extends CommonPropertyGridProps {
   model: IPropertyGridModel;
   eventHandler: IPropertyGridEventHandler;
-  highlight?: HighlightingComponentProps & { filteredTypes?: FilteredType[] };
+  highlight?: HighlightingComponentProps & {
+    filteredTypes?: FilteredType[];
+  };
+  propertyCategoryRendererManager?: PropertyCategoryRendererManager;
```
Comment (grigasp) [inline ui/components/src/ui-components/propertygrid/component/VirtualizedPropertyGrid.tsx:48]: Should be `@alpha`?

**Review by grigasp — CHANGES_REQUESTED** [review]
- Docs need to be updated for this new category renderer attribute.
- Would be good to add a `NextVersion.md` topic with an example regarding this.
- Whoever is going to use custom renderers through presentation rules, there will be a need to find out what ECInstance's content needs to be rendered inside the custom category. Do you have an idea how that would be done?

**Comment by roluk** [comment]
>* Whoever is going to use custom renderers through presentation rules, there will be a need to find out what ECInstance's content needs to be rendered inside the custom category. Do you have an idea how that would be done?

Custom renderers are able to access category contents through `props.categoryItem.getChildren()`, but I might have misunderstood the question.

File: presentation/common/src/presentation-common/rules/content/modifiers/PropertyCategorySpecification.ts [hunk 3]
```diff
@@ -30,4 +32,7 @@ export interface PropertyCategorySpecification {
 
   /** Should this category be auto-expanded. Defaults to `false`. */
   autoExpand?: boolean;
+
+  /** Custom category renderer specification. */
+  renderer?: PropertyRendererSpecification;
```
[the remark this answers is not in the document]
Comment (roluk) [reply presentation/common/src/presentation-common/rules/content/modifiers/PropertyCategorySpecification.ts]: I chose to go the other route and unify both types as `CustomRendererSpecification` in cc53605efa5a799cc01e7afe95b380fc5574967c.

File: ui/components/src/ui-components/propertygrid/component/VirtualizedPropertyGrid.tsx (hunk 2 above)
Comment (roluk) [reply ui/components/src/ui-components/propertygrid/component/VirtualizedPropertyGrid.tsx:50]: Thanks, added in 190c6ba68b154d50acb35e81494009364051dab5.

File: ui/components/src/ui-components/propertygrid/component/VirtualizedPropertyGridWithDataProvider.tsx (hunk 1 above)
Comment (roluk) [reply ui/components/src/ui-components/propertygrid/component/VirtualizedPropertyGridWithDataProvider.tsx:31]: Thanks, added in 190c6ba68b154d50acb35e81494009364051dab5.

File: presentation/common/src/presentation-common/rules/content/modifiers/PropertyCategorySpecification.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (grigasp) [reply presentation/common/src/presentation-common/rules/content/modifiers/PropertyCategorySpecification.ts]: looks good, thanks

**Comment by grigasp** [comment]
> > * Whoever is going to use custom renderers through presentation rules, there will be a need to find out what ECInstance's content needs to be rendered inside the custom category. Do you have an idea how that would be done?
> 
> Custom renderers are able to access category contents through `props.categoryItem.getChildren()`, but I might have misunderstood the question.

The problem that we're trying to solve is:
- we're displaying properties for some element A
- the element has related properties on class B
- the consumer wants not only custom rendering of B properties, but also query some additional data related to the B instance and render that.

With your changes we provide a custom renderer which has access to `PropertyRecords` through `props.categoryItem.getChildren()`. From there we can find a `PropertyRecord` that we want, but I don't see how to get `InstanceKey` of the element whose data the record contains - it is essential for building a query for additional data.  The only way to find out the information is through the data provider that produced the record, so we have to make it available to the renderer in some way.

**Review by pmconne — APPROVED** [review]
Approving NextVersion.md changes.

**Review by grigasp — APPROVED** [review]

**Review by DanEastBentley — APPROVED** [review]

**Pull request merged**